R/peaks-functions.R

```diff
@@ -562,3 +562,50 @@ joinPeaksGnps <- function(x, y, xPrecursorMz = NA_real_,
         x[idx, , drop = FALSE]
     }
 }
+
+#' Combine peaks within each peak matrix if the difference of their m/z is
+#' smaller than `tolerance` and `ppm`. Peak grouping is performed with the
+#' `MsCoreUtils::group` function.
+#'
+#' Eventually present peak annotations are dropped (replaced by `NA`) for peaks
+#' that are combined.
+#'
+#' Note that `weighted = TRUE` overrides `mzFun` using the `weighted.mean`
+#' to calculate the aggregated m/z values.
+#'
+#' @author Johannes Rainer
+#'
+#' @importFrom stats weighted.mean
+#'
+#' @noRd
+.peaks_combine <- function(x, ppm = 20, tolerance = 0,
+                           intensityFun = base::mean, mzFun = base::mean,
+                           weighted = TRUE, spectrumMsLevel,
+                           msLevel = spectrumMsLevel, ...) {
+    if (!spectrumMsLevel %in% msLevel || !length(x))
+        return(x)
+    grps <- group(x[, "mz"], tolerance = tolerance, ppm = ppm)
+    lg <- length(grps)
+    if (grps[lg] == lg)
+        return(x)
+    mzs <- split(x[, "mz"], grps)
+    mrg <- lengths(mzs) > 1L
+    ints <- split(x[, "intensity"], grps)
+    if (weighted)
+        mzs <- unlist(mapply(
+            mzs, ints, FUN = function(m, i) weighted.mean(m, i + 1),
+            SIMPLIFY = FALSE, USE.NAMES = FALSE), use.names = FALSE)
+    else mzs <- vapply1d(mzs, mzFun)
+    ints <- vapply1d(ints, intensityFun)
+    if (ncol(x) > 2) {
+        idx <- match(seq_len(grps[lg]), grps)
+        lst <- lapply(x[!colnames(x) %in% c("mz", "intensity")], function(z) {
+            z <- z[idx]
+            z[mrg] <- NA
+            z
```

Comment:
My comment above might not be correct... you always set the additional peaks variables to 0 upon combining, even if they are constant? Is this meant to stay like that, or do you consider updating in the future. Imagine for example the situation where you want to combine groups of peaks based on one of these additional peaks variable - dropping would be a shame.